src/js/azdata/azdataQueryPlan.js

```diff
@@ -1,3 +1,126 @@
+class Point {
+    constructor(x, y) {
+        this.x = x;
+        this.y = y;
+    }
+}
+class GraphNodeLayoutHelper {
+    constructor() {
+        this.layoutPoints = [];
+    }
+
+    checkInvariant() {
+        var last = new Point(0, 0);
+
+        for (var i = 0; i < this.layoutPoints.length; i++) {
+            if (last.x > this.layoutPoints[i].x || last.Y > this.layoutPoints[i].y) {
+                console.log('Graph layout failed.');
+                // do not throw exception, in case of failure we render graph incorrectly 
+            }
+
+            last = this.layoutPoints[i];
+        }
+    }
+
+    updateNodeLayout(xPosition, yPosition) {
+        this.checkInvariant();
+
+        // First cover edge cases
+
+        // Empty list
+        if (this.layoutPoints.length === 0) {
+            this.layoutPoints.push(new Point(xPosition, yPosition));
+            return;
+        }
+
+        // Single Element
+        if (this.layoutPoints.length === 1) {
+            if (xPosition < this.layoutPoints[0].x) {
+                this.layoutPoints.splice(0, 0, new Point(xPosition, yPosition));
+            }
+            else if (xPosition === this.layoutPoints[0].x) {
+                this.layoutPoints[0] = new Point(this.layoutPoints[0].x, Math.max(this.layoutPoints[0].y, yPosition));
```

Comment:
It seems unnecessary to instantiate a new `Point` when the current one can just have its y-position updated.